pkg/jobs/collector.go

```diff
@@ -0,0 +1,181 @@
+package jobs
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/aquasecurity/trivy-kubernetes/pkg/k8s"
+	batchv1 "k8s.io/api/batch/v1"
+	v1 "k8s.io/api/core/v1"
+	k8sapierror "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	NodeCollectorName = "node-collector"
+	TrivyNamespace    = "trivy-system"
+)
+
+type Collector interface {
+	ApplyAndCollect(ctx context.Context, nodeName string) (string, error)
+	Apply(ctx context.Context, nodeName string) (*batchv1.Job, error)
+	Cleanup(ctx context.Context)
+}
+
+type jobCollector struct {
+	cluster k8s.Cluster
+	// timeout duration for collection job to complete it task before is cancelled default 0
+	timeout        time.Duration
+	logsReader     LogsReader
+	labels         map[string]string
+	annotation     map[string]string
+	templateName   string
+	namespace      string
+	name           string
+	serviceAccount string
+}
+
+type CollectorOption func(*jobCollector)
+
+func WithTimetout(timeout time.Duration) CollectorOption {
+	return func(jc *jobCollector) {
+		jc.timeout = timeout
+	}
+}
+
+func WithJobLabels(labels map[string]string) CollectorOption {
+	return func(jc *jobCollector) {
+		jc.labels = labels
+	}
+}
+
+func WithJoAnnotation(annotation map[string]string) CollectorOption {
+	return func(jc *jobCollector) {
+		jc.annotation = annotation
+	}
+}
+
+func WithJobNamespace(namespace string) CollectorOption {
+	return func(jc *jobCollector) {
+		jc.namespace = namespace
+	}
+}
+
+func WithName(name string) CollectorOption {
+	return func(jc *jobCollector) {
+		jc.name = name
+	}
+}
+
+func WithServiceAccount(sa string) CollectorOption {
+	return func(jc *jobCollector) {
+		jc.serviceAccount = sa
+	}
+}
+
+func WithJobTemplateName(name string) CollectorOption {
+	return func(jc *jobCollector) {
+		jc.templateName = name
+	}
+}
+
+func NewCollector(
+	cluster k8s.Cluster,
+	opts ...CollectorOption,
+) Collector {
+	jc := &jobCollector{
+		cluster:    cluster,
+		timeout:    0,
+		logsReader: NewLogsReader(cluster.GetK8sClientSet()),
+	}
+	for _, opt := range opts {
+		opt(jc)
+	}
+	return jc
+}
+
+// ApplyAndCollect deploy k8s job by template to  specific node  and namespace, it read pod logs
+// cleaning up job and returning it output (for cli use-case)
+func (jb *jobCollector) ApplyAndCollect(ctx context.Context, nodeName string) (string, error) {
+	job, err := GetJob(
+		WithTemplate(jb.templateName),
+		WithNodeSelector(nodeName),
+		WithNamespace(TrivyNamespace),
+		WithJobServiceAccount(jb.serviceAccount),
+		WithJobName(fmt.Sprintf("%s-%s", jb.templateName, nodeName)))
+	if err != nil {
+		return "", fmt.Errorf("running node-collector job: %w", err)
+	}
+
+	_, err = jb.getTrivyNamespace(ctx)
+	if err != nil {
+		if k8sapierror.IsNotFound(err) {
+			trivyNamespace := &v1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: TrivyNamespace}}
```

Comment:
Why do we create the namespace for `ApplyAndCollect`, but not for `Apply`? `Apply` also creates a job.

Reply:
`ApplyAndCollect` is for trivy and  `Apply` is for operator , so no need to create namespace for operator